torch/amp/grad_scaler.py

```diff
@@ -0,0 +1,567 @@
+from collections import defaultdict, abc
+from enum import Enum
+from typing import Any, Dict, List, Optional, Tuple, cast
+import inspect
+import warnings
+
+import torch
+
+
+__all__ = ["OptState", "GradScaler"]
+
+class _MultiDeviceReplicator:
+    """
+    Lazily serves copies of a tensor to requested devices.  Copies are cached per-device.
+    """
+    def __init__(self, master_tensor: torch.Tensor) -> None:
+        self.master = master_tensor
```

Comment:
I see you removed the assertion for the device check = CUDA and XLA. We will still want to keep these during the refactoring until we can confidently say all devices are supported.